README.md

```diff
@@ -1,16 +1,1831 @@
+        }
+      }
+    }
+  },
+  "roots": [
+    {
+      "/": "bafyreic7ypuk4xd4kvintknoxfgy33ttkr3kv2vp2fenpylecofhuswiye"
+    }
+  ]
+}
+```
+
+By analogy, above examples can be interpreted roughly as follows:
+
+```js
+const createDraft = crud.create("https://example.com/blog/posts", {
+  payload: {
+    title: "How UCAN Tasks Changed My Life",
+    body: "This is the story of how one spec changed everything...",
+  },
+})
+
+const getEditors = crud.read("https://example.com/users/editors")
+
+const notify = msg.send("mailto:akiko@example.com", {
+  to: (await createDraft).ok,
+  subject: "Coffee",
+  body: (await getEditors).ok,
+})
+```
+
+While a [Await] MAY be substituted for any field in a [Task], substituting the `do` field is NOT RECOMMENDED. The `do` field is critical in understanding what kind of action will be performed, and schedulers SHOULD use this fields to grant atomicity, parallelize tasks, and so on.
+
+After resolution, the [Invocation] MUST be validated against the [Authorization] and linked UCAN proofs by the [Executor]. A Promise resolved to an [Invocation] that is not backed by a valid UCAN MUST NOT be executed, and SHOULD return an unauthorized error to the user. A Promise resolved to an [Invocation] with the [Authorization] that does not include invoked [Task] MUST NOT be executed, and SHOULD return an unauthorized error to the user.
+
+Promises MAY be used across [Invocation]s with a same [Authorization], or across [Invocation]s with different [Authorization] and MAY even be across multiple Invokers and Executors. As long as the invocation can be resolved, it MAY be promised. This is sometimes referred to as ["promise pipelining"](http://erights.org/elib/distrib/pipeline.html).
+
+## 9.2 Promise
+
+A `Promise` has a same representation as [Invocation] with only difference being that `auth` field is OPTIONAL. Promise MAY be used by an [Invocation] to [Await] another [Invocation] authorized with the same [Authorization]. [Invocation] MAY be used directly everywhere else in place of the [Promise].
+
+### 9.2.1 Schema
+
+```ipldsch
+# Promise is an Invocation with optional 'auth' field which if omitted
+# is implicitly an 'auth' of the Invocation that contains Await.
+type Promise struct {
+  v       SemVer
+
+  run     &Task
+  # Receipt of the invocation that caused this invocation
+  cause   optional &Invocation
+  # Task authorization. If omitted can be interpreted as requires
+  # authorization 
+  auth    optional &Authorization
+
+  meta    {String : any}
+
+  prf     [&UCAN]
+}
+
+```
+
+## 9.3 Await
+
+An `Await` describes an output of the [Invocation]. An `Await` MUST resolve to an output [Result] with `await/*` variant. If unwrapping success or failure case is desired, corresponding `await/ok` or `await/error` variants MUST be used.
+
+An [Invocation] from the [Await] is resolved by:
+
+1. Creating a new [Invocation].
```

Comment:
It creates a new Invocation? Can you expand on this?

Reply:
I think some of the WEB specs use this kind of language, what I mean is if `Promise` has `auth` omitted it is implicitly an `auth` of the task awaiting it, in other words

```js
resolve = (promise, parentInvocation) =>
  promise.auth ? promise : { ...promise, auth: parentInvocation.auth }
```

Not sure what the best way to describe this in plain English

Reply:
Interesting 🤔 I have a lot of use cases for this kind of thing, but was going to nest it to make it really clear where the auth is coming from, but I guess this also work (though it's maybe less clear). Does DAG House have a use case for this too, or is it to try to satisfy IPVM requirements?

Reply:
Perhaps to clarify my above comment: if it's for your use case I'm fine leaving it in for now. If it's for mine, we can probably remove this section.

Reply:
My point is we need it do accomplish what I've described in https://github.com/ucan-wg/invocation/pull/10/files#r1100959759 otherwise you won't be able to use same authorization.

I think we'll be totally fine with following schema instead

```ipldsch
type Await union {
  | &Invocation "await/*"
  | &Invocation "await/ok"
  | &Invocation "await/error"
} representation keyed
```

Which in following example 

```mermaid
flowchart BR
    update-dns("with: dns:example.com?TYPE=TXT
                do: crud/update")
    notify-bob("with: mailto://alice@example.com
                do: msg/send
                to: bob@example.com")
    notify-carol("with: mailto://alice@example.com
                  do: msg/send
                  to: carol@example.com")

    log-as-done("with: https://example.com/report
                do: crud/update")

    update-dns --> notify-bob --> log-as-done
    update-dns --> notify-carol --> log-as-done
```

Would imply producing 3 different authorizations:

- 1. That has `updateDNS` in it's scope
- 2. That has `notifyBob` and `notifyCarol` in scope.
- 3. That has `report` in scope.

With currently described resolution logic all four could be authorized together, in fact that is how generated example does it now https://observablehq.com/@gozala/ucan-invocation-spec#batchedPipelineExample

Reply:
To answer your question if you're ok with schema change and consequently increased number of signatures needed, I'm fine with that. However if we do want to allow use of one signature / auth in the above case we really need spec to explain what the implied `auth` of the awaited invocation is, but maybe it could be described better somehow.

I suspect we would just sign each invocation individually to avoid complexity (which I got a feeling of making that observable), so I'd say it's really up to you.

Reply:
Reading this https://github.com/ucan-wg/invocation/pull/10/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R813 I am realizing we can't get rid of this otherwise it would be impossible to express dependencies among tasks in the effect.

However I'm sensing that perhaps what you have problem with is that I prescribe what `auth` of awaited `Promise` is implied to be. I'm guessing it is ok for IPVM because you don't really care about the `auth` since you'll look for the receipt by the `Task` CID anyway.

But given that we key by the Invocation, we need to resolve the Invocation CID from the Promise and that is why I felt a need to describe exactly how you get an Invocation CID. 

While it may make sense for us to consider keying things by Task CID instead, it does still bother me that spec would prescribe that much, ideally this would be an implementation detail.

As aside keying by Task CID seems troublesome in non pure setting like ours, it seems really difficult to guarantee that same task will produce same receipt regardless who authorized it or set of proofs they've provided.